build_catalog.sh

```diff
@@ -1,105 +1,186 @@
+                removed_versions="$version $removed_versions"
+            fi
+        done
+        [[ -n "$removed_versions" ]] && log "The following versions will be removed: $removed_versions"
+        cp "$new_bundle_versions_file" "$bundle_versions_file"
     fi
-    log "Found previous bundle version $prev_version"
-  popd
-  rm -rf $tmp_dir
-  docker rm tmp_$$
+    prev_operator_version=$(tail -n 1 "$bundle_versions_file")
+else
+    log "No $bundle_versions_file exist. This is the first time the operator is built"
+    new_bundle_versions_file="$bundle_versions_file"
+    touch "$bundle_versions_file"
 fi
 
-# Build/push the new bundle
-pushd deploy/bundle
-  log "Creating bundle $BUNDLE_IMAGE:$current_commit"
-  if [[ $prev_version != "" ]]; then
-    export REPLACE_VERSION=$prev_version
-  fi
-  export BUNDLE_IMAGE_TAG=$current_commit
-  export OPERATOR_IMAGE_TAG=v$version
-  export VERSION=$version
-  make bundle
-  docker tag $BUNDLE_IMAGE:$current_commit $BUNDLE_IMAGE:latest
-
-  log "Pushing the bundle $BUNDLE_IMAGE:$current_commit to repository"
-  $docker_cmd push $BUNDLE_IMAGE:$current_commit
-  # Do not push the latest tag here.  If there is a problem creating the catalog then
-  # pushing the latest tag here will mean subsequent runs will be extracting a bundle
-  # version that isn't referenced in the catalog.  This will result in all future
-  # catalog creation failing to be created.
-popd
-
-# Download opm build
-curl -L https://github.com/operator-framework/operator-registry/releases/download/v$opm_version/linux-amd64-opm -o ./opm
-chmod u+x ./opm
-
-# Create/push a new catalog via opm
-log "Pulling existing latest catalog $CATALOG_IMAGE"
-$docker_cmd pull $CATALOG_IMAGE:latest && exists=1 || exists=0
-if [ $exists -eq 1 ]; then
-  from_arg="--from-index $CATALOG_IMAGE:latest"
+if [[ "$OPERATOR_VERSION" == "$prev_operator_version" ]]; then
+    log "stopping script as $OPERATOR_VERSION version was already built, so no need to rebuild it"
+    exit 0
 fi
 
-if [[ "$from_arg" == "" ]]; then
-  log "Creating new catalog $CATALOG_IMAGE"
-else
-  log "Updating existing catalog $CATALOG_IMAGE"
+log "Creating bundle image $BUNDLE_IMAGE:$CURRENT_COMMIT"
+
+mkdir -p "$MANIFEST_DIR"
+template=$(mktemp -p "$temp_dir")
+./"$BUNDLE_DEPLOY_DIR"/generate-csv-template.py > "$template"
+oc process --local -o yaml --raw=true \
+    IMAGE="$OPERATOR_IMAGE" \
+    IMAGE_TAG="$OPERATOR_IMAGE_TAG" \
+    VERSION="$OPERATOR_VERSION" \
+    REPLACE_VERSION="$prev_operator_version" \
+    -f "$template" > "$CSV"
+
+if [[ "$prev_operator_version" == "" ]]; then \
+    sed -i.bak "/replaces/d" "$CSV"
+    rm -f "$CSV.bak"
+fi
+
+$CONTAINER_ENGINE build -t "$BUNDLE_IMAGE:$CURRENT_COMMIT" "$BUNDLE_DEPLOY_DIR"
+$CONTAINER_ENGINE tag "$BUNDLE_IMAGE:$CURRENT_COMMIT" "$BUNDLE_IMAGE:latest"
+
+# opm won't get anything locally, so we need to push the bundle
+log "Pushing $BUNDLE_IMAGE:$CURRENT_COMMIT"
+$engine_cmd push "$BUNDLE_IMAGE:$CURRENT_COMMIT"
```

Comment:
`DRY_RUN` condition here?

Reply:
This is actually an interesting problem.  If you don't push the bundle image then you can't know if the catalog will be created correctly because the opm always pulls the image from the registry and doesn't use local images even if the one it wants is already on the system.  If we could find a way to use local images for catalog creation, then definitely this should have DRY_RUN condition.  If we can't, maybe delete the pushed image in DRY_RUN?  Ugly though.

Reply:
oh, right. Yeah, that's icky. I would say the image is harmless, except that if OLM tries to deploy with it, it'll fail because it refers to a bundle that we _didn't_ push :(

Another (also icky) option would be to use a different org/repo for DRY_RUN images...

Reply:
OLM won't try to deploy the bundle unless it is part of the catalog.  If the catalog doesn't get pushed in DRY_RUN then we should be ok.  Not ideal, but at least it wouldn't get deployed and subsequent bundle builds should push new versions of that bundle tag until one is functional to add to the catalog.

We could push to quay stage for dry_run maybe?  My concern there is we would have to introspect the CATALOG/BUNDLE image information passed in and change the repo location.

Reply:
Actually, what about just appending -dryrun to the end of the bundle/catalog image names for DRY_RUN mode?  We'd be pushing different image names to quay than the production names so no change of overlap

Reply:
+1.

I don't remember (or possibly never knew) how OLM decides which catalog image to use for deployment. How does it know to avoid `-dryrun`?

Reply:
https://github.com/app-sre/deployment-validation-operator/blob/master/deploy/openshift/deployment-validation-operator-olm.yaml

^^ Those pieces are what tie everything together.  The catalogsource is the piece that ties OLM to a specific catalog, which provides the available bundles.

Reply:
...and that file is what app-interface's saas config points to, populates, and deploys. Got it. Thanks.

Reply:
Rob is right. I try to explain with the comment that I needed this even in dry-run mode. I will look into implementing the different tag in dry-run mode, it is a good idea.

Reply:
Ah yes, <tag>-dryrun is probably better than <IMAGE>-dryrun.  I like that better.

Reply:
Fixed in 4cf9aec